### DISCUSSION TITLE
Autoscaling Runner Scale Sets mode is now generally available

### DISCUSSION THREAD

**Discussion by Link-** [actions/actions-runner-controller#2775] [Announcements]
Last January we introduced [the autoscaling runner scale sets mode](https://github.com/actions/actions-runner-controller/discussions/2174) to actions runner controller (ARC) and this mode is now generally available. We published the announcement in the [GitHub changelog](https://github.blog/changelog/2023-06-30-github-actions-actions-runner-general-availability/) end of June, and we're excited to share the news here as well.

## A few things to clarify

If you're not familiar with this new mode, I highly encourage you to either [read our documentation](https://gh.io/arc-docs). If you want a more guided alternative, you can watch this [deep dive video](https://www.youtube.com/watch?v=_F5ocPrv6io) by yours truly.

### Maintenance and support

The autoscaling runner scale sets mode is now generally available, which means that GitHub will provide maintenance and support for this mode. We recommend that you create tickets with our support team for assistance.

Issues created in this repository will be triaged, and we will address them based on priority. We urge you to read our documentation and put all the effort possible to try to resolve the issues by yourselves. We have extensive testing pipelines, and we're observing the behaviour of this product across different environments. When a wide-spread issue is identified, we will address it as soon as possible. If you face a problem, more likely than not, it's something related to the specifics of your environment.

Please don't make us spend time on issues that could have been avoided by a quick read of the documentation or a quick search in the issues / discussions. This will take away from the time that could have been spent addressing more important problems. We cannot provide any guarantees towards response times for issues created in this repository.

The following modes and assets are supported by GitHub:

- [Autoscaling runner scale sets mode](https://github.com/actions/actions-runner-controller/tree/master/controllers/actions.github.com)
- All the [resources belonding to: `apis/actions.github.com/v1alpha1`](https://github.com/actions/actions-runner-controller/tree/master/apis/actions.github.com/v1alpha1)
- ARC [CI and release workflows](https://github.com/actions/actions-runner-controller/tree/master/.github/workflows)
- The [new runner image](https://github.com/actions/runner/pkgs/container/actions-runner)
- The gha-runner-scale-set-* helm charts: [controller](https://github.com/actions/actions-runner-controller/tree/master/charts/gha-runner-scale-set-controller) and [runnner scale set](https://github.com/actions/actions-runner-controller/tree/master/charts/gha-runner-scale-set)
- ARC documentation: [gh.io/arc-docs](https://gh.io/arc-docs)

Everything else, including but not limited to:

- [Pull driven scaling, and Webhook driven scaling modes](https://github.com/actions/actions-runner-controller/tree/master/controllers/actions.summerwind.net)
- All the [resources belonging to the `actions.summerwind.net/v1alpha1`](https://github.com/actions/actions-runner-controller/tree/master/apis/actions.summerwind.net/v1alpha1)
- The [actions-runner-controller helm charts](https://github.com/actions/actions-runner-controller/tree/master/charts/actions-runner-controller)
- The [old runner images](https://github.com/actions/actions-runner-controller/tree/master/runner):
  - [actions-runner](https://github.com/actions-runner-controller/actions-runner-controller/pkgs/container/actions-runner-controller%2Factions-runner)
  - [actions-runner-dind-rootless](https://github.com/actions-runner-controller/actions-runner-controller/pkgs/container/actions-runner-controller%2Factions-runner-dind-rootless)
  - [actions-runner-dind](https://github.com/actions-runner-controller/actions-runner-controller/pkgs/container/actions-runner-controller%2Factions-runner-dind)
- Pre-migration [actions-runner-controller](https://github.com/actions-runner-controller/) organization

Will continue to be maintained **by the community only** led by @mumoshu. GitHub will not provide any maintenance or support for these assets.

### Releases

We will continue to release new versions of the autoscaling runner scale sets mode, and we will continue to publish the new runner image with every runner binaries release. We will also continue to publish the helm charts for the controller and the runner scale set.

The helm charts for the autoscaling runner scale sets mode **will only be published as OCI packages**. We believe this is the best way to distribute helm charts, and we will not publish them as tarballs via GitHub pages.

- The helm chart and controller-manager images are available here: <https://github.com/orgs/actions/packages?repo_name=actions-runner-controller>
- The runner image is available here: <https://github.com/actions/runner/pkgs/container/actions-runner>

All legacy assets are published to the [actions-runner-controller organization](https://github.com/orgs/actions-runner-controller/packages).

## What's next?

The release `gha-runner-scale-set-0.5.0` is currently our highest priority. This release will include the following changes: <https://github.com/actions/actions-runner-controller/milestone/11>.

We do not have any plans to do a `1.x.x` release at the moment. Even though we are generally available, we don't consider our APIs or helm charts stable enough to be considered `1.x.x`. We will continue to release `0.x.x` versions until we feel confident that we can provide stable and backward compatible APIs and helm charts.

## Contributions

We highly appreciate all the community's contributions to ARC. Please continue creating issues and discussions, and we will continue to triage them and address them based on priority. We're working on new contributions guidelines to manage the process of accepting contributions. We will share them as soon as they're ready.

Until then, **if you plan to spend effort and time on a substantial change (anything that's larger than 10 lines of code), please open an issue and discuss it with us first**. We offer no guarantees that your changes will be accepted.

Thank you all for all your trust and support ❤️ and thank you to countless contributors who help make ARC awesome!

**Comment by simpson** [comment]
ubuntu image yassssss

**Comment by maxdanilov** [comment]
That's a perfect summary; thank you! Our team has been guessing in the last months about the future of this project and the dynamics between the community and GitHub-developed components - and this post answers these questions perfectly.

I'd like to suggest the following:
- link this discussion in the [readme](https://github.com/actions/actions-runner-controller) - otherwise, anyone starting with the project won't know the important points and distinctions you mention here (e.g., which controller to use?)
- the link to `read our documentation` is broken; I assume it's intended to lead [here](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners-with-actions-runner-controller/quickstart-for-actions-runner-controller)

**Reply by Link-** [reply]
Thank you @maxdanilov - the broken links have been fixed and the README should be updated soon as we gradually phase out the old docs.

**Comment by maxdanilov** [comment]
@mumoshu are there plans to develop the old ARC past 0.27? (0.28, 0.29, etc.)? Or it will only receive the maintenance patches for 0.27.x?

**Reply by mumoshu** [reply]
Hey! I'll try to keep releasing new versions for the old ARC on demand @maxdanilov 
Just curious, though- what could be the motivation for keep using the old ARC for you?!

**Reply by maxdanilov** [reply]
@mumoshu The main motivation is that the new Runner Scale Set does not look mature enough compared to the classic ARC (e.g. the metrics are still not there: https://github.com/actions/actions-runner-controller/pull/2559), and GitHub is not bridging the gap quickly, for example, the latest 0.4.0 release happened over 4 months ago, which is not agile enough for such a new developing project.

I don't expect new features from ARC 0.27.x, but I would really appreciate it if at least the new release is cut from the latest `master` every now and then. The latest version is 0.27.4 released over 3 months ago, and because of the lack of recent releases, it's not possible to use newer stability fixes like this one: https://github.com/actions/actions-runner-controller/pull/2671. We rely on this PR to add stability to our scalers but have to build our own Docker image from `master` which is suboptimal.

Thank you for your work on developing and supporting this project, it's highly appreciated 🙇

**Comment by AlonAvrahami** [comment]
Thank you @Link- Great news!!
Can you tell if there is a support installing the  actions_metrics_server with this new mode?
I think it would be awesome to be able to collect metrics of the runners and the workflows/jobs  :)

**Reply by Link-** [reply]
Metrics are coming as part of the gha-runner-scale-set-0.5.0 release

**Comment by hamishforbes** [comment]
Do you intend to support the all-in-one / `dockerdWithinRunnerContainer` feature from the legacy ARC?

It's quite tricky to properly set resource requests/limits when some jobs will use resources on the 'machine' (i.e. the `runner` container) and others, which specify a container in the workflow, will use resources on the `dind` container.

I don't really want to use the kubernetes mode either as this has other drawbacks.

**Comment by dm3ch** [comment]
@Link- Is it possible to add ability to have multiple labels on runnerSet?
https://github.com/actions/actions-runner-controller/issues/2445

I'm interested if Github api endpoint for runnerset able to recieve multiple labels? I would like to try to create PR with supporting multiple labels if maintainers have no time to work on that  and if github api able to handle multiple labels.

P.S. I don't even understand which label types are possible

**Reply by Link-** [reply]
Labels are not supported and will not be supported for runner scale sets. Please read the docs:

- https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners-with-actions-runner-controller/about-actions-runner-controller#using-arc-runners-in-a-workflow

**Comment by Link-** [comment]
I'm locking this conversation. For support and/or feature requests kindly open other discussions or issues.
